File: 0744-network-delay-time/0744-network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        graph = collections.defaultdict(list) # 이중 그래프를 만들고
        
        for time in times:
            graph[time[0]].append((time[2], time[1]))
        
        pq = []
        costs = {}
        
        heapq.heappush(pq, (0, k))

        while pq:
            cur_v, cur_node = heapq.heappop(pq)
            if cur_node not in costs:
                costs[cur_node] = cur_v
                for next_cost, next_node in graph[cur_node]:
                    sum_cost = cur_v + next_cost
                    heapq.heappush(pq, (sum_cost, next_node))
        
        # 최소 비용중에 최고 최소비용 (signal을 보낼 수 있는 최소 비용)
        for i in range(1, n+1):
            if i not in costs:
                return -1
        
        return max(costs.values())

File: 0744-network-delay-time/0744-network-delay-time.kt
class Solution {
    fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
        val graph = List(n+1) { mutableListOf<Pair<Int, Int>>() }

        for (time in times) {
            graph[time[0]].add(time[2] to time[1])
        }

        val costs: MutableMap<Int, Int> = mutableMapOf()
        val pq: PriorityQueue<Pair<Int, Int>> = PriorityQueue<Pair<Int, Int>>(compareBy {it.first})
        pq.add(0 to k)

        while (pq.isNotEmpty()) {
            val (curCost, curNode) = pq.poll()
            if (curNode !in costs) {
                costs[curNode] = curCost
                for ((nextCost, nextNode) in graph[curNode]) {
                    val sum: Int = curCost + nextCost
                    pq.add(sum to nextNode)
                }
            }
        }

        return if (costs.size == n) costs.values.maxOrNull() ?: -1
        else -1
    }
}